### PR TITLE
CBL-4052: objc API ends with error updated with NoThrow

### PR DESCRIPTION
**Pull request by jayahariv**
* No throw for objc APIs ends with error
* Make it consistent for all APIs instead of do..try..catch in save, delete functions

### CODE REVIEW

File: Objective-C/CBLDatabase.h [hunk 1]
```diff
@@ -111,7 +111,7 @@ __deprecated_msg("Use [database defaultCollection].count instead.");
  @return The CBLDocument object.
  */
 - (nullable CBLDocument*) documentWithID: (NSString*)id
-__deprecated_msg("Use [database defaultCollection] documentWithID:] instead.");
+__deprecated_msg("Use [database defaultCollection] documentWithID:] instead.") NS_SWIFT_NOTHROW;
```
Comment (pasin) [inline Objective-C/CBLDatabase.h:114]: Does it need here as well when no out error?

File: Objective-C/CBLDatabase.h [hunk 2]
```diff
@@ -543,7 +543,7 @@ __deprecated_msg("Use [[database defaultCollection] getDocumentExpirationWithID:
  */
 - (nullable CBLCollection*) createCollectionWithName: (NSString*)name
                                                scope: (nullable NSString*)scope
-                                               error: (NSError**)error;
+                                               error: (NSError**)error NS_SWIFT_NOTHROW;
```
Comment (pasin) [inline Objective-C/CBLDatabase.h:546]: Is there a case that a null collection returned without an error?

File: Objective-C/CBLDatabase.h [hunk 3]
```diff
@@ -581,7 +581,7 @@ __deprecated_msg("Use [[database defaultCollection] getDocumentExpirationWithID:
  @param error On return, the error if any.
  @return Default Scope, or nil if an error occurred.
  */
-- (nullable CBLScope*) defaultScope: (NSError**)error;
+- (nullable CBLScope*) defaultScope: (NSError**)error NS_SWIFT_NOTHROW;
```
Comment (pasin) [inline Objective-C/CBLDatabase.h:584]: I don't think there is a case that null is returned without an error.

File: Objective-C/CBLDatabase.h [hunk 4]
```diff
@@ -590,7 +590,7 @@ __deprecated_msg("Use [[database defaultCollection] getDocumentExpirationWithID:
         default collection is deleted.
  @return Default collection, or nil if an error occurred.
  */
-- (nullable CBLCollection*) defaultCollection: (NSError**)error;
+- (nullable CBLCollection*) defaultCollection: (NSError**)error NS_SWIFT_NOTHROW;
```
Comment (pasin) [inline Objective-C/CBLDatabase.h:593]: There is no case as of now to return a null default collection without an error.

File: Objective-C/CBLQueryFactory.h [hunk 5]
```diff
@@ -33,7 +33,7 @@ NS_ASSUME_NONNULL_BEGIN
  @param error On return, the error if any., the given query string is invalid(e.g., syntax error).
  @return query created using the given expression string, or nil if an error occurred.
  */
-- (nullable CBLQuery*) createQuery: (NSString*)query error: (NSError**)error;
+- (nullable CBLQuery*) createQuery: (NSString*)query error: (NSError**)error NS_SWIFT_NOTHROW;
```
Comment (pasin) [inline Objective-C/CBLQueryFactory.h:36]: I don't think there is a case that a null is returned without error.

File: Objective-C/CBLReplicator.h [hunk 6]
```diff
@@ -169,7 +169,7 @@ __deprecated_msg("Use [replicator pendingDocumentIDsForCollection:error:] instea
  @return A  set of document Ids, each of which has one or more pending revisions. If error, nil.
  */
 - (nullable NSSet<NSString*>*) pendingDocumentIDsForCollection: (CBLCollection*)collection
-                                                         error: (NSError**)error;
+                                                         error: (NSError**)error NS_SWIFT_NOTHROW;
```
Comment (pasin) [inline Objective-C/CBLReplicator.h:172]: Is there a case that null is returned without an error?

File: Objective-C/CBLScope.h [hunk 7]
```diff
@@ -53,7 +53,7 @@ extern NSString* const kCBLDefaultScopeName;
  @param error On return, the error if any.
  @return Collections in the scope, or nil if an error occurred.
  */
-- (nullable NSArray<CBLCollection*>*) collections: (NSError**)error;
+- (nullable NSArray<CBLCollection*>*) collections: (NSError**)error NS_SWIFT_NOTHROW;
```
Comment (pasin) [inline Objective-C/CBLScope.h:56]: I think there is no case to return null collections without error.

**Review by pasin — COMMENTED** [review]
I think this PR is risky right be helium and I don't see it necessary as well.

**Comment by jayahariv** [comment]
This PR is necessary! 

Example: 
```- (nullable Collection) collection: (NSError**)error ```

This ObjC Api will generate swift API with a non-optional return type. 

```func collection() throws -> Collection ```

So If the ObjC API returns nil in some case, it will cause swift to receive an exception/error.

File: Objective-C/CBLDatabase.h (hunk 1 above)
Comment (jayahariv) [reply Objective-C/CBLDatabase.h:114]: its not necessary, I might have accidentally updated with collection documentWithID:error:

File: Objective-C/CBLQueryFactory.h (hunk 5 above)
Comment (pasin) [inline Objective-C/CBLQueryFactory.h:36]: No cases that null will be returned without an error?

File: Objective-C/CBLReplicator.h (hunk 6 above)
Comment (pasin) [inline Objective-C/CBLReplicator.h:172]: No cases that null will be returned without an error?

File: Objective-C/CBLScope.h (hunk 7 above)
Comment (pasin) [inline Objective-C/CBLScope.h:56]: No cases that null will be returned without an error?

File: Objective-C/CBLDatabase.h [hunk 8]
```diff
@@ -530,7 +530,7 @@ __deprecated_msg("Use [[database defaultCollection] getDocumentExpirationWithID:
  @return list of collections in the scope, or nil if an error occurred.
  */
 - (nullable NSArray<CBLCollection*>*) collections: (nullable NSString*)scope
-                                            error: (NSError**)error;
+                                            error: (NSError**)error NS_SWIFT_NOTHROW;
```
Comment (pasin) [inline Objective-C/CBLDatabase.h:533]: No cases that null will be returned without an error?

**Review by pasin — CHANGES_REQUESTED** [review]
Basically, please change the PR to add NS_SWIFT_NOTHROW only on the methods that need to return null with null error.

File: Objective-C/CBLDatabase.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (jayahariv) [reply Objective-C/CBLDatabase.h:546]: I have made consistent with other call, that's why.

**Review by pasin — APPROVED** [review]

**Pull request merged**